File: 0238-product-of-array-except-self/0238-product-of-array-except-self.js
/**
 * @param {number[]} nums
 * @return {number[]}
 */
var productExceptSelf = function(nums) {
    let n = nums.length;
    let answer = new Array(n);
    let left = new Array(n);
    let right = new Array(n);

    left[0] = 1;
    for (let i = 1; i < n; i++) {
        left[i] = nums[i - 1] * left[i - 1];
    }

    right[n - 1] = 1;
    for (let i = n - 2; i >= 0; i--) {
        right[i] = nums[i + 1] * right[i + 1];
    }

    for (let i = 0; i < n; i++) {
        answer[i] = left[i] * right[i];
    }

    return answer;
};

File: 0238-product-of-array-except-self/0238-product-of-array-except-self.py
class Solution:
    def productExceptSelf(self, nums: List[int]) -> List[int]:
        n = len(nums)
        prefix_products = [1] * n
        suffix_products = [1] * n  
        answer = [1] * n
        
        for i in range(1, n):
            prefix_products[i] = prefix_products[i - 1] * nums[i - 1]

        for i in range(n - 2, -1, -1):
            suffix_products[i] = suffix_products[i + 1] * nums[i + 1]

        for i in range(n):
            answer[i] = prefix_products[i] * suffix_products[i]

        return answer